### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by IraCherginyshka**

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+const checkToken = (req, res, next) => {
+  if (req.header('Authorization').split(' ')[1] === userData[req.header('Name')]) {
+    return next();
+  }
+  return errorHandler(res);
+};
+
+app.post('/sign_up', (req, res) => {
+  const user = new User(req.body);
```
Comment (oleksandrstarov) [inline server/app.js]: no variable is required

File: server/app.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        return res.status(200).json({
+          token,
+          user,
+        });
+      }
+      return errorHandler(res);
+    })
+    .catch(() => {
+      return errorHandler(res);
+    });
```
Comment (oleksandrstarov) [inline server/app.js]: should be moved to db.js, db.createUser(req.body).then().catch()

File: src/app/app.module.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import { BrowserModule } from '@angular/platform-browser';
+import { NgModule } from '@angular/core';
+import { ReactiveFormsModule } from '@angular/forms';
+import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
+import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
+
+import { ToastContainerModule, ToastrModule } from 'ngx-toastr';
```
Comment (oleksandrstarov) [inline src/app/app.module.ts]: move one line up

File: src/app/components/game-board/game-board.component.scss [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+.board-canvas {
+  border: $primary-border;
+  background-color: $background-color;
+}
+
+.overlay {
+  position: absolute;
+  width: $board-width + 2px;
+  height: $board-height + 3px;
+  background-color: rgba($primary-color, 0.85);
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.scss]: add variable to .85

File: src/app/components/game-board/game-board.component.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+  ngOnDestroy(): void {
+    this.subscriptionState.unsubscribe();
+    this.subscriptionMove.unsubscribe();
+    this.subscriptionNext.unsubscribe();
+    this.subscriptionLevel.unsubscribe();
+  }
+
+  private rotateFigure(figureMatrix: FiguresColors[][]): FiguresColors[][] {
+    const reverseMatrix = [...figureMatrix];
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: [...figureMatrix].reverse();

File: src/app/components/game-board/game-board.component.ts [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    reverseMatrix.reverse();
+    return reverseMatrix[0].map((item, index) => reverseMatrix.map((line) => line[index]));
+  }
+
+  private redrawBoard(): void {
+    const newFigure = new FigureModel();
+    const newBoard = new BoardModel(this.ctx, false);
+
+    this.currentMatrix = newFigure.showFigure(
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts:200]: newFigure >> new FigureModel()

File: src/app/components/game-board/game-board.component.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+  private setInitialBoardState(): void {
+    this.lineWithFigure = 0;
+    this.figurePosition = CENTRAL_ITEM;
+    this.duration = DELAY_DEFAULT;
+  }
+
+  private playGame(): void {
+    this.timeInterval = window.setInterval(() => {
+      if (this.checkCollisionDetection(0, this.currentFigure)) {
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: add const hasCollision = this.checkCollisionDetection(0, this.currentFigure);

File: src/app/components/game-board/game-board.component.ts [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        this.redrawBoard();
+        this.lineWithFigure += 1;
+      } else if (
+        !this.checkCollisionDetection(0, this.currentFigure) &&
+        this.lineWithFigure === 0
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: !lineWith...

File: src/app/components/game-board/game-board.component.ts [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+  }
+
+  private resetGame(): void {
+    this.boardMatrix = BoardModel.makeBoardEmptyMatrix(
+      QUANTITY_BLOCKS_WIDTH,
+      QUANTITY_BLOCKS_HEIGHT,
+    );
+    this.gameService.updateFigures();
+    this.redrawBoard();
+    clearInterval(this.timeInterval);
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: stopGame()

File: src/app/components/game-board/game-board.component.ts [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+
+  private lostGame(): void {
+    this.isLostGame = true;
+    this.isPlaying = false;
+    this.gameService.setLostGame().subscribe(() => {
+      this.toastrService.warning('You have successfully added your result to Leaderboard');
+    });
+    this.gameService.updateFigures();
+    this.setInitialBoardState();
+    clearInterval(this.timeInterval);
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: stopGame()

File: src/app/components/game-controls/game-controls.component.ts [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+@Component({
+  selector: 'atg-game-controls',
+  templateUrl: './game-controls.component.html',
+  styleUrls: ['./game-controls.component.scss'],
+})
+export class GameControlsComponent implements OnInit, OnDestroy {
+  public moveStep = FiguresMovement;
+
+  private isPlaying: boolean;
```
Comment (oleksandrstarov) [inline src/app/components/game-controls/game-controls.component.ts:16]: separate public/private

File: src/app/components/game-header/game-header.component.html [hunk 12]
```diff
@@ -0,0 +1,16 @@
+<div class="header">
+  <div class="logo">tetris</div>
+  <atg-player-profile *ngIf="userIsAuthenticated; else login"></atg-player-profile>
+  <div class="wrap-links">
+    <nav class="bar-links container">
+      <a class="header-link" [routerLink]="['/login']">login</a>
```
Comment (oleksandrstarov) [inline src/app/components/game-header/game-header.component.html]: please check routerLinkActive

File: src/app/components/game-header/game-header.component.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    this.authSubscription = this.userService
+      .getAuthListener()
+      .subscribe((user: PlayerData | null) => {
+        if (user) {
+          this.toastrService.success('Welcome to AwesomeTetrisGame!');
+          this.userIsAuthenticated = true;
+        } else {
+          this.userIsAuthenticated = false;
```
Comment (oleksandrstarov) [inline src/app/components/game-header/game-header.component.ts]: can be this.userIsAuthenticated = !!user

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]

File: src/app/components/game-header/game-header.component.html (hunk 12 above)
Comment (IraCherginyshka) [reply src/app/components/game-header/game-header.component.html]: Thanks

**Pull request closed without merging**